image-decoders code fails to compile with -Wthread-safety-analysis

litherum opened the pull request:
#### 121ab62cae80a9f4b937c0cc4c3633ae52502dad
<pre>
image-decoders code fails to compile with -Wthread-safety-analysis
<a href="https://bugs.webkit.org/show_bug.cgi?id=253397">https://bugs.webkit.org/show_bug.cgi?id=253397</a>
rdar://106239523

Reviewed by Simon Fraser.

This patch mostly adds WTF_REQUIRES_LOCK() annotations to places that need it.

It also switches out use of C++ API to C API, for binary compatibility.

No test because there is no behavior change.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::imageOutCallback):
(WebCore::JPEGXLImageDecoder::prepareColorTransform):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.h:
* Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderRAII.h: Added.
(JxlDecoderPtr::JxlDecoderPtr):
(JxlDecoderPtr::operator=):
(JxlDecoderPtr::~JxlDecoderPtr):
(JxlDecoderPtr::reset):
(JxlDecoderPtr::operator bool const):
(JxlDecoderPtr::get const):
(JxlDecoderMake):

Canonical link: <a href="https://commits.webkit.org/261240@main">https://commits.webkit.org/261240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65f964e5c917c50a293a01637963c3402b63d2df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110985 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20124 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43552 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2413 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119832 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114944 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11226 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2071 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116738 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99151 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103471 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97928 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30815 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44405 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12642 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32151 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/86314 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13187 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9133 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18602 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7798 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15135 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderRAII.h:
```diff
@@ -0,0 +1,90 @@
+#include <algorithm>
+#include <jxl/decode.h>
+
+class JxlDecoderPtr {
```
Why does the file name not match the class name?

litherum replied:
Good idea!

unicodon reviewed (approved):
LGTM 👍

litherum committed:
image-decoders code fails to compile with -Wthread-safety-analysis
https://bugs.webkit.org/show_bug.cgi?id=253397
rdar://106239523

Reviewed by Simon Fraser.

This patch mostly adds WTF_REQUIRES_LOCK() annotations to places that need it.

It also switches out use of C++ API to C API, for binary compatibility.

No test because there is no behavior change.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.cpp:
(WebCore::JPEGXLImageDecoder::imageOutCallback):
(WebCore::JPEGXLImageDecoder::prepareColorTransform):
* Source/WebCore/platform/image-decoders/jpegxl/JPEGXLImageDecoder.h:
* Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderRAII.h: Added.
(JxlDecoderPtr::JxlDecoderPtr):
(JxlDecoderPtr::operator=):
(JxlDecoderPtr::~JxlDecoderPtr):
(JxlDecoderPtr::reset):
(JxlDecoderPtr::operator bool const):
(JxlDecoderPtr::get const):
(JxlDecoderMake):

Canonical link: https://commits.webkit.org/261240@main

aproskuryakov commented on Source/WebCore/platform/image-decoders/jpegxl/JxlDecoderPtr.h:
```diff
@@ -0,0 +1,90 @@
+/*
+* Copyright (C) 2021 Apple Inc. All rights reserved.
```
2021?
